quantity.hpp

```diff
@@ -406,10 +406,12 @@ class quantity {
  */
 template<typename Q1, typename Q2> using mul_quantity_t = quantity< mul_unit_t<typename Q1::unit_type, typename Q2::unit_type>
                                                                   , std::ratio_multiply<typename Q1::scale_type, typename Q2::scale_type>
-                                                                  , decltype(typename Q1::value_type{}*typename Q2::value_type{}) >;
+                                                                  , decltype(typename Q1::value_type{}*typename Q2::value_type{})
+                                                                  , typename std::enable_if<std::is_same<typename Q1::tag_type, typename Q2::tag_type>::value, typename Q1::tag_type>::type>;
 template<typename Q1, typename Q2> using div_quantity_t = quantity< div_unit_t<typename Q1::unit_type, typename Q2::unit_type>
                                                                   , std::ratio_divide<typename Q1::scale_type, typename Q2::scale_type>
-                                                                  , decltype(typename Q1::value_type{}/typename Q2::value_type{}) >;
+                                                                  , decltype(typename Q1::value_type{}/typename Q2::value_type{})
+                                                                  , typename std::enable_if<std::is_same<typename Q1::tag_type, typename Q2::tag_type>::value, typename Q1::tag_type>::type>;
```

Comment:
IIUC, this would invalidate all multiplications and divisions where the tags are not the same. I am not sure this is a good idea. For example, this wouldn't compile anymore: 
 ```cpp
auto reactive_energy = 10._kVAr * 2_h;
```

While this might help in a specific case, I am pretty sure this is not what users expect in general.

Reply:
After some more consideration: 

* The above example would not currently compile, since a **`static_assert`** in **`operator*`** would kick in.
* Even if the static_assert wasn't there, it would not do what is expected: Since `mul_quantity_t` currently always returns a unit tagged with **`void`**, the result would not be reactive energy, but untagged (active) energy.

However, the above example *should* compile and result in energy with the reactive tag. The problem is to find a strategy which allows this without allowing things users do not expect. A strategy to deal with this problem might be:

1. If all quantities involved have the same tag, keep that tag.
2. If the tags involved are **`void`** and one other tag, tag the result with other tag. 
3. If more than one non-**`void`** tag is involved, demand explicit casting.

Reply:
_After verbal discussions:_ 

The issue at hand here is a ratio equation where an object of a tagged quantity (_reactive power_) is multiplied with an object of the same type and then divided by an object of the same type. The expected behavior of this would be that the result is an object of the same type. However, with the current handling of tags in multiplication and division the result is of the same quantity _with the tag stripped_. 

Unfortunately, while the above solution would help with this issue, it would not solve other issues which users would assume to work. For example, when multiplying _reactive power_ (power with an `reactive` tag) with _time_, and dividing the result (_reactive energy_) by _reactive power_, the result would be _reactive time_, which is nonsense. The correct action here would be to cancel out the reactive flag in the division.

Doing this in general, however, would invalidate the result of the above mentioned ratio equation, which would result in a quantity with the tag stripped from it.